lib/opentelemetry/trace/exporter/circuit.lua

```diff
@@ -0,0 +1,77 @@
+local util = require("opentelemetry.util")
+local otel_global =  require("opentelemetry.global")
+
+local _M = {
+    OPEN = "open",
+    CLOSED = "closed",
+    HALF_OPEN = "half-open"
+}
+
+local mt = {
+    __index = _M
+}
+
+
+function _M.new(options)
+    options = options or {}
+    local self = {
+        circuit_reset_timeout_ms = options["circuit_reset_timeout_ms"] or 5000,
+        failure_threshold = options["failure_threshold"] or 5,
+        failure_count = 0,
+        open_start_time_ms = nil,
+        state = "closed",
+    }
+    return setmetatable(self, mt)
+end
+
+function _M.should_make_request(self)
+    if self.state == self.CLOSED then
+        return true
+    end
+
+    if self.state == self.OPEN then
+        if (util.gettimeofday_ms() - self.open_start_time_ms) > self.circuit_reset_timeout_ms then
+            self.state = self.HALF_OPEN
+            self.open_start_time_ms = nil
+        else
+            return false
+        end
+    end
+
+    if self.state == self.HALF_OPEN then
+        return true
+    end
+
+    ngx.log(ngx.ERR, "should not have gotten here")
+    return false
+end
+
+function _M.process_failed_request(self)
+    self.failure_count = self.failure_count + 1
+
+    if self.state == self.CLOSED and self.failure_count >= self.failure_threshold then
+        otel_global.metrics_reporter:add_to_counter("otel.bsp.circuit_breaker_opened", 1)
+        self.state = self.OPEN
+        self.open_start_time_ms = util.gettimeofday_ms()
+    end
+
+    if self.state == self.HALF_OPEN then
+        otel_global.metrics_reporter:add_to_counter("otel.bsp.circuit_breaker_opened", 1)
+        self.state = self.OPEN
+        self.open_start_time_ms = util.gettimeofday_ms()
+    end
+end
+
+function _M.process_succeeded_request(self)
+    if self.state == self.CLOSED then
+        return
+    end
+
+    if self.state == self.HALF_OPEN then
+        otel_global.metrics_reporter:add_to_counter("otel.bsp.circuit_breaker_closed", 1)
+        self.state = self.CLOSED
```

Comment:
self.failure_count = 0

Reply:
I do not think we care about failure_count when in half-open state. We only care about failure count when in closed state. I reset that here 43d2b717deab94689520e9a0ff1d220d30195f3b.

Reply:
👍  43d2b717deab94689520e9a0ff1d220d30195f3b